### PR TITLE
Add nomv command 

### LINKED ISSUE #5082: [RFC] Add nomv command in PPL
## Proposal — Command syntax, options, and precise semantics

Command
`nomv <field>`

### field (required)
The name of a multivalue field whose values will be converted into a single value.

### Constraints
1. Must be a direct field reference.
2. Must be a multivalue (array) field.
3. If the field does not exist in the current schema, the command fails. 

## Precise semantics 
nomv operates on the current result set produced by the pipeline. The command is row-local and does not perform grouping or aggregation

## Per-row transformation

For each input row:
Read the value of <field>. If the value is a multivalue (array), convert it into a single scalar string. Replace <field> with the converted scalar value. Preserve all other fields unchanged. 

## Value rendering
Array elements are joined using a newline delimiter ("\n"). The delimiter is fixed and not configurable. 

## Ordering
1. Value order in the resulting string follows the existing order of elements in the array at the point where nomv executes.
2. If deterministic ordering is required, users must ensure ordering before nomv by controlling how the multivalue field is constructed. 

## Missing field (schema-level)
The command fails with an error if <field> does not exist in the current schema.

## Null field value (row-level)
If <field> exists but its value is null for a given row, the output value remains null.

## Empty array
If <field> is an empty array, the output value is an empty string ("").

## Null elements inside array
Null elements are ignored and do not produce delimiters in the output.

## Type transformation
Input type: ARRAY<T>
Output type: STRING

## Error handling
The command fails if:
1. The target field does not exist in the current schema.
2. The target is not a direct field reference.
3. The target field is not a multivalue (array) type. 

## 1. Basic
Input:

| user | tags        |
| joe  | ["a","b"]   |
| sam  | ["x"]       |

Query: 
... | nomv tags

| user | tags  |
| joe  | a\nb  |
| sam  | x     |

## 2. Multiple rows (row-local behavior)
Input: 
| user | tags            |
| joe  | ["a","b"]       |
| joe  | ["c","d"]       |

Query: 
... | nomv tags

Output: 

| user | tags  |
| joe  | a\nb  |
| joe  | c\nd  |

##  3. Non-consecutive rows (no grouping) 
Input: 

| user | tags      |
| joe  | ["a"]     |
| sam  | ["x"]     |
| joe  | ["b"]     |

Query: 
... | nomv tags

Output: 
| user | tags |
| joe  | a    |
| sam  | x    |
| joe  | b    |

##  4. Empty array
Input:

| user | tags |
| joe  | []   |

Query: 
... | nomv tags

Output: 
| user | tags |
| joe  |      |

## 5. Null elements inside array
Input:

| user | tags              |
| joe  | ["a", null, "b"]  |

Query: 
... | nomv tags

Output: 
| user | tags |
| joe  | a\nb |

## 6. Null field value
Input:

| user | tags |
| joe  | null |

Query:

... | nomv tags

Output:
| user | tags |
| joe  | null |

##  7. Missing target field
Input:

| user | action |
| joe  | login  |

Query:
... | nomv tags

Result:
Error: field [tags] not found in schema

##  8. Scalar field (invalid target)
Input:

| user | status |
| joe  | ok     |

Query:
... | nomv status

Result:
Error: field [status] is not a multivalue field

 ## 9: Array with all nulls
  Input:
  | user | tags         |
  | joe  | [null, null] |

  Query: ... | nomv tags

  Output:
  | user | tags |
  | joe  | ""   |

## 10: Array with empty strings
  Input:
  | user | tags           |
  | joe  | ["a", "", "b"] |

  Query: ... | nomv tags

  Output:
  | user | tags   |
  | joe  | "a\n\nb" |

---

##  Approach
Implement nomv as a streaming projection command. No grouping, buffering, or state is required. The command rewrites the logical plan by replacing <field> with a scalar expression derived from the array value.

## Relationship to existing commands
1. nomv is not an aggregation command.
2. nomv does not modify row cardinality.
3. nomv is a per-row transformation and can be modeled as a projection.

### PR DESCRIPTION
**Pull request by srikanthpadakanti**
### Description
Converts values of the specified multivalue field into one single value. Separates the values using a new line "\n delimiter.  

### Related Issues
Resolves #5082
https://github.com/opensearch-project/sql/issues/5082

### Check List
- [ X ] New functionality includes testing.
- [ X ] New functionality has been documented.
 - [ X ] New functionality has javadoc added.
 - [ X ] New functionality has a user manual doc added.
- [ X ] New PPL command [checklist](https://github.com/opensearch-project/sql/blob/main/docs/dev/ppl-commands.md) all confirmed.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ X ] Commits are signed per the DCO using `--signoff` or `-s`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/analysis/Analyzer.java [hunk 1]
```diff
@@ -541,6 +542,11 @@ public LogicalPlan visitMvCombine(MvCombine node, AnalysisContext context) {
     throw getOnlyForCalciteException("mvcombine");
   }
 
+  @Override
+  public LogicalPlan visitNoMv(NoMv node, AnalysisContext context) {
+    throw getOnlyForCalciteException("nomv");
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply core/src/main/java/org/opensearch/sql/analysis/Analyzer.java:553]: Following the existing code template - where I do not see javadocs.

File: core/src/main/java/org/opensearch/sql/ast/AbstractNodeVisitor.java [hunk 2]
```diff
@@ -471,4 +472,8 @@ public T visitAddColTotals(AddColTotals node, C context) {
   public T visitMvCombine(MvCombine node, C context) {
     return visitChildren(node, context);
   }
+
+  public T visitNoMv(NoMv node, C context) {
+    return visitChildren(node, context);
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply core/src/main/java/org/opensearch/sql/ast/AbstractNodeVisitor.java:482]: Following the existing code template - where I do not see javadocs.

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRelNodeVisitor.java [hunk 3]
```diff
@@ -3339,6 +3340,11 @@ private void restoreColumnOrderAfterArrayAgg(
     relBuilder.project(projections, projectionNames, /* force= */ true);
   }
 
+  @Override
+  public RelNode visitNoMv(NoMv node, CalcitePlanContext context) {
+    return visitEval((Eval) node.rewriteAsEval(), context);
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply core/src/main/java/org/opensearch/sql/calcite/CalciteRelNodeVisitor.java:3312]: done

File: core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java [hunk 4]
```diff
@@ -78,6 +78,7 @@ public enum BuiltinFunctionName {
   MVAPPEND(FunctionName.of("mvappend")),
   INTERNAL_APPEND(FunctionName.of("append"), true),
   MVJOIN(FunctionName.of("mvjoin")),
+  NOMV(FunctionName.of("nomv")),
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply core/src/main/java/org/opensearch/sql/expression/function/BuiltinFunctionName.java]: done.

File: docs/user/ppl/cmd/nomv.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+`nomv` is a transforming command: it modifies the specified field without changing the number of rows in the result set.
+
+### Key behaviors
+
+- The field must be of **ARRAY type**. For scalar fields, use the `array()` function to create an array first.
+- The specified field is **replaced** with a string containing all array elements joined by newline (`\n`) characters.
+- **NULL values within the array are automatically filtered out** before joining.
+- If the field doesn't exist, an error is returned.
+- The operation uses Calcite's ARRAY_JOIN function internally (same underlying implementation as mvjoin).
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: done.

File: docs/user/ppl/cmd/nomv.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+## Example 2: nomv with multiple fields
+
+```ppl
+source=accounts
+| where account_number=1
+| eval location = array(city, state)
+| nomv location
+| fields account_number, location
+```
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: done.

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            + " | eval arr = array('a', 'b', 'c') | nomv arr | head 1 | fields arr";
+
+    JSONObject result = executeQuery(q);
+
+    verifySchema(result, schema("arr", null, "string"));
+
+    verifyDataRows(result, rows("a\nb\nc"));
+
+    Assertions.assertEquals(1, result.getJSONArray("schema").length());
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java:177]: done

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            + "      LogicalTableScan(table=[[scott, EMP]])\n";
+    verifyLogical(root, expectedLogical);
+
+    String expectedSparkSql =
+        "SELECT ARRAY_JOIN(ARRAY('a', 'b'), '\n') `arr`, UPPER(ARRAY_JOIN(ARRAY('a', 'b'),"
+            + " '\n')) `arr_upper`\n"
+            + "FROM `scott`.`EMP`\n"
+            + "LIMIT 1";
+    verifyPPLToSparkSQL(root, expectedSparkSql);
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java:401]: done.

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    verifyLogical(root, expectedLogical);
+
+    String expectedSparkSql =
+        "SELECT ARRAY_JOIN(ARRAY('web', 'production', 'east'), '\n') `arr`"
+            + LS
+            + "FROM `scott`.`EMP`"
+            + LS
+            + "LIMIT 1";
+    verifyPPLToSparkSQL(root, expectedSparkSql);
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java:72]: done.

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+  public void testNoMvNonExistentField() {
+    String ppl = "source=EMP | eval arr = array('a', 'b') | nomv does_not_exist | head 1";
+
+    Exception ex = assertThrows(Exception.class, () -> getRelNode(ppl));
+
+    String msg = String.valueOf(ex.getMessage());
+    org.junit.Assert.assertTrue(
+        "Expected error message to mention missing field. Actual: " + msg,
+        msg.toLowerCase().contains("does_not_exist") || msg.toLowerCase().contains("field"));
+  }
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java:204]: done

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+        "LogicalProject(EMPNO=[$0], arr=[$8])\n"
+            + "  LogicalSort(fetch=[1])\n"
+            + "    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4],"
+            + " SAL=[$5], COMM=[$6], DEPTNO=[$7], arr=[ARRAY_JOIN(array(), '\n')])\n"
+            + "      LogicalTableScan(table=[[scott, EMP]])\n";
+    verifyLogical(root, expectedLogical);
+
+    String expectedSparkSql =
+        "SELECT `EMPNO`, ARRAY_JOIN(ARRAY(), '\n') `arr`\n" + "FROM `scott`.`EMP`\n" + "LIMIT 1";
+    verifyPPLToSparkSQL(root, expectedSparkSql);
```
[the remark this answers is not in the document]
Comment (srikanthpadakanti) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLNoMvTest.java:340]: done

File: docs/user/ppl/cmd/nomv.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+```ppl
+source=accounts
+| nomv does_not_exist
+```
+
+Expected output:
+```text
+{'reason': 'Invalid Query', 'details': 'Field [does_not_exist] not found.', 'type': 'IllegalArgumentException'}
+Error: Query returned no data
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: what is it? is it response from rest API?

File: docs/user/ppl/cmd/nomv.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+---
+
+## Notes
+
+- The `nomv` command is only available when the Calcite query engine is enabled.
+- This command is particularly useful when you need to export or display multivalue fields as single strings.
+- The newline delimiter (`\n`) is fixed and cannot be customized. For custom delimiters, use the `mvjoin` function directly in an eval expression.
+- NULL values are automatically filtered out during the join operation, so they do not contribute empty strings to the output.
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: what does "join operation" means, could u be more specific?

File: docs/user/ppl/cmd/nomv.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+## Related commands
+
+- `mvjoin()` -- Function used by nomv internally to join array elements with a custom delimiter
+- [`eval`](eval.md) -- Create computed fields using the `array()` and `mvjoin()` functions
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: How it releated to eval?

File: docs/user/ppl/cmd/nomv.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+fetched rows / total rows = 1/1
++----------------+----------+
+| account_number | location |
+|----------------+----------|
+| 1              | Brogan   |
+|                | IL       |
++----------------+----------+
+```
+
+## Example 3: Error when field does not exist
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: nit, it is standard behaviour of all command, we can remove it.

File: docs/user/ppl/cmd/nomv.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+## Description
+
+The `nomv` command converts a multivalue (array) field into a single-value string field by joining all array elements with newline characters (`\n`). This operation is performed in-place, replacing the original field with its joined string representation.
+
+`nomv` is a transforming command: it modifies the specified field without changing the number of rows in the result set.
+
+### Key behaviors
+
+- The field must be a **direct field reference** of **ARRAY type**. For scalar fields, use the `array()` function to create an array first.
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: The field must be ARRAY type.

File: docs/user/ppl/cmd/nomv.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+`nomv` is a transforming command: it modifies the specified field without changing the number of rows in the result set.
+
+### Key behaviors
+
+- The field must be a **direct field reference** of **ARRAY type**. For scalar fields, use the `array()` function to create an array first. 
+- The specified field is **replaced** with a string containing all array elements joined by newline (`\n`) characters.
+- **NULL values within the array are automatically filtered out** before joining.
+- If the field doesn't exist, an error is returned.
+- The operation uses Calcite's ARRAY_JOIN function internally (same underlying implementation as mvjoin).
```
Comment (penghuo) [inline docs/user/ppl/cmd/nomv.md]: Remove it. This doc is for end user.

File: docs/user/ppl/index.md [hunk 18]
```diff
@@ -78,11 +78,12 @@ source=accounts
 | [describe command](cmd/describe.md) | 2.1 | stable (since 2.1) | Query the metadata of an index. |
 | [explain command](cmd/explain.md) | 3.1 | stable (since 3.1) | Explain the plan of query. |
 | [show datasources command](cmd/showdatasources.md) | 2.4 | stable (since 2.4) | Query datasources configured in the PPL engine. |
-| [addtotals command](cmd/addtotals.md) | 3.5 | stable (since 3.5) | Adds row and column values and appends a totals column and row. | 
+| [addtotals command](cmd/addtotals.md) | 3.5 | stable (since 3.5) | Adds row and column values and appends a totals column and row. |
 | [addcoltotals command](cmd/addcoltotals.md) | 3.5 | stable (since 3.5) | Adds column values and appends a totals row. |
 | [transpose command](cmd/transpose.md) | 3.5 | stable (since 3.5) | Transpose rows to columns. |
 | [mvcombine command](cmd/mvcombine.md) | 3.5 | stable (since 3.4) | Combines values of a specified field across rows identical on all other fields. |
-  
+| [nomv command](cmd/nomv.md) | 3.6 | experimental (since 3.6) | Converts a multivalue field to a single-value string by joining elements with newlines. |
```
Comment (penghuo) [inline docs/user/ppl/index.md]: why experimental? this command is simple.

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+
+    JSONObject result = executeQuery(q);
+
+    verifySchema(result, schema("arr", null, "string"));
+
+    verifyDataRows(result, rows(""));
+  }
+
+  @Test
+  public void testNoMvArrayWithNullValues() throws IOException {
```
Comment (penghuo) [inline integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java]: Evidence: The array is `array('first', 'second', 'third')` — zero null elements. This test does not exercise the null-element handling required by issue #5082 Example 5 (`["a", null, "b"]` → `"a\nb"`).
Next step: Rename to `testNoMvArrayWithThreeElements` and add a separate test that constructs an array with a null element (e.g. via a nullable field) and asserts the null is skipped in the joined output.

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+    int status = ex.getResponse().getStatusLine().getStatusCode();
+
+    Assertions.assertEquals(400, status, "Unexpected status. ex=" + ex.getMessage());
+
+    String msg = ex.getMessage();
+    Assertions.assertTrue(
+        msg.contains("does_not_exist") || msg.contains("field") || msg.contains("Field"), msg);
+  }
+}
```
Comment (penghuo) [inline integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java:359]: Evidence: No test covers the case where the array field itself is null for a row. Issue #5082 Example 6 specifies the output should remain null, not empty string or error.
Next step: Add a test that produces a null array field value (e.g. a field that is not assigned for some rows) and asserts the nomv output is null.

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+
+    JSONObject result = executeQuery(q);
+
+    verifySchema(result, schema("arr", null, "string"));
+
+    verifyDataRows(result, rows("single"));
+  }
+
+  @Test
+  public void testNoMvEmptyArray() throws IOException {
```
Comment (penghuo) [inline integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java]: Evidence: `testNoMvEmptyArray` covers `array()` → `""`, but issue #5082 Example 9 specifies a distinct case: an array where all elements are null should also produce `""`. No test covers this.
Next step: Add `testNoMvAllNullElementsArray` that constructs an array of all-null elements (via nullable fields) and asserts the output is `""`.

File: docs/user/ppl/cmd/nomv.md (hunk 12 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: The format shown is from the PPL CLI tool (not direct REST API) but the other documentation using bash/curl examples should show the JSON format.

File: docs/user/ppl/cmd/nomv.md (hunk 13 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: Elaborated: 
NULL values within the array are automatically filtered out when converting the array to a string, so they do not appear in the output or contribute empty lines.

File: docs/user/ppl/cmd/nomv.md (hunk 14 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: The relationship to eval is:
  1. You must use eval with array() function to create an array field before you can use nomv on it
  2. Alternatively, you can use eval with mvjoin() function directly if you need a custom delimiter instead of using nomv 
  

if you feel that is misleading - I can change it to 
**"- `array()` -- Function to create arrays from multiple fields (used in `eval` command)** 

if that is TMI and misleading. I can delete that.

File: docs/user/ppl/cmd/nomv.md (hunk 15 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: done.

File: docs/user/ppl/cmd/nomv.md (hunk 16 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: addressed.

File: docs/user/ppl/cmd/nomv.md (hunk 17 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/cmd/nomv.md]: Sure.

File: docs/user/ppl/index.md (hunk 18 above)
Comment (srikanthpadakanti) [reply docs/user/ppl/index.md]: done.

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java (hunk 19 above)
Comment (srikanthpadakanti) [reply integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java]: understood. Addressed this by adding the necessary logic in ArrayFunctionImpl

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java (hunk 20 above)
Comment (srikanthpadakanti) [reply integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java:359]: understood. Addressed this by adding the necessary logic in ArrayFunctionImpl

File: integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java (hunk 21 above)
Comment (srikanthpadakanti) [reply integ-test/src/test/java/org/opensearch/sql/calcite/remote/CalciteNoMvCommandIT.java]: understood. Addressed this by adding the necessary logic in ArrayFunctionImpl
